### ISSUE TITLE
Cursor is confused in example of "play-on-repeat"

### ISSUE THREAD

**Issue by watersoft123** [paulrosen/abcjs#658]
During playback, if each messure contains 4 beats, the playback is normal. If I add a messure with 2 or 3 beats(this is often the case in many music scores), the end or start position of the cursor is confused.
![20210407120119](https://user-images.githubusercontent.com/57562864/113809217-3da65200-979a-11eb-8a8a-1498faa233e2.jpg)

**Comment by paulrosen** [comment]
I think that is just in that example - the start position is determined with
```
var start = (startMeasure.value - 1) / NUM_MEASURES;
```
You can factor in the extra measures there if you want. The same with the end which is calculated  in the `onBeat` callback.

**Comment by watersoft123** [comment]
> I think that is just in that example - the start position is determined with
> 
> ```
> var start = (startMeasure.value - 1) / NUM_MEASURES;
> ```
> 
> You can factor in the extra measures there if you want. The same with the end which is calculated in the `onBeat` callback.

I add extraMeasure and change  start & end, the cursor position of end is ok,but  the still a problem with the starting position.

`timingCallbacks = new ABCJS.TimingCallbacks(visualObj, {
		extraMeasuresAtBeginning:1/4,          //add extraMeasure
		beatCallback: cursorControl.onBeat,
		eventCallback: cursorControl.onEvent
});
cursorControl.onStart();
		synth.prime().then(function () {
		var start = ((startMeasure.value-1) * NUM_BEATS_PER_MEASURE+1) / (4*NUM_MEASURES);    //change start
		synth.seek(start);
		timingCallbacks.setProgress(start);
		synth.start();
		timingCallbacks.start();
 });
.......
.......
self.onBeat = function(beatNumber, totalBeats, totalTime) {
	var end = (endMeasure.value-1) * NUM_BEATS_PER_MEASURE+1;    //change end
	if (beatNumber >= end) {
		var start = ((startMeasure.value-1) * NUM_BEATS_PER_MEASURE+1) / (4*NUM_MEASURES);    //change start
		synth.seek(start);
		timingCallbacks.setProgress(start);
	}
};`
![20210410201009](https://user-images.githubusercontent.com/57562864/114269674-ed99ea80-9a3a-11eb-9e5f-8c6c4b15905d.png)

**Comment by watersoft123** [comment]
One of my thoughts is whether it is possible to click on a note to return a number as the starting position, and then click on another as the ending position.

**Comment by watersoft123** [comment]
I have another question. I downloaded a soundfont file and placed it in a folder called sound, but I don't know how to reference this sf2 file. I refer to the help file and add one of the following options, but none of them play sound. I don't know where the error is?
  soundFontUrl: "sound/FluidR3_GM.sf2"
  soundFontUrl: "file:///sound/FluidR3_GM.sf2"
  soundFontUrl: "file:///sound/"
  soundFontUrl: "http://sound/FluidR3_GM.sf2"
  soundFontUrl: "http://sound/"
![0210410204259](https://user-images.githubusercontent.com/57562864/114270148-60a46080-9a3d-11eb-97ee-2d1e7e163f31.png)

**Comment by ehsansh** [comment]
@watersoft123 
About starting music from clicked note. You can use `seek` function but it is not accurate. Paul told me he would update it https://github.com/paulrosen/abcjs/issues/640#issuecomment-803363087

About the other question.
If you mean sf2 format. Abcjs does not support it yet. But it is possible to add some mp3 files in that folder with one of midi standard instrument names. You can read [here](https://github.com/paulrosen/abcjs/issues/643).

I always search my questions in the issues. I find some useful information from previous issues. You can do the same.

**Comment by watersoft123** [comment]
Thanks ehsansh. Let us wait for Paul's good news  of seek function. 
about soundfont. If abcjs does not support sf2 format, what format does it default?

https://paulrosen.github.io/abcjs/overview/faq.html#for-node
In above website，There is a reply on how to use local soundfont, but I can’t open the link named sandbox. 
"How do I use the sound fonts I have saved locally in my ABCJS project?
See the sandbox to see the syntax"

**Comment by watersoft123** [comment]
Instead of sf2 file,I download acoustic_grand_piano-mp3. js and volin-mp3.js.  I set "%%MIDI program 40", But still can’t play sound offline.

**Comment by ehsansh** [comment]
What you have done is correct but have you checked the `soundFontUrl` ? Do you see any error?
You can use the network tab of browser to see whether your audio files are correctly downloaded and buffered.
In order to use abcjs offline I downloaded  FluidR3_GM.sf2 from [here](http://gleitz.github.io/midi-js-soundfonts/)
Then I used these codes:
```js
var midiBuffer = new ABCJS.synth.CreateSynth();
	const audioParams = { 
		options: { 
			soundFontUrl: '/soundfonts/FluidR3_GM'
		}
	};
midiBuffer.init({
	audioContext: new AudioContext(),
	visualObj: visualObj,					
	options: {
		soundFontUrl: '/soundfonts/FluidR3_GM'
	}
})
```
And it works. Using the `%%MIDI program` in the abc string changes the instrument.

**Comment by watersoft123** [comment]
> Paul told me he does not support sf2 format I don't know why he uses this word while the audio files are just mp3, ogg,etc.
> What you have done is correct but have you checked the `soundFontUrl` ? Do you see any error?
> You can use the network tab of browser to see whether your audio files are correctly downloaded and buffered.
> In order to use abcjs offline I downloaded FluidR3_GM.sf2 from [here](http://gleitz.github.io/midi-js-soundfonts/)
> Then I used these codes:
> 
> ```js
> var midiBuffer = new ABCJS.synth.CreateSynth();
> 	const audioParams = { 
> 		options: { 
> 			soundFontUrl: '/soundfonts/FluidR3_GM'
> 		}
> 	};
> midiBuffer.init({
> 	audioContext: new AudioContext(),
> 	visualObj: visualObj,					
> 	options: {
> 		soundFontUrl: '/soundfonts/FluidR3_GM'
> 	}
> }).then(function (response) {
> 					// console.log(response);
> 
> 	if (synthControl) {
> 		synthControl.setTune(visualObj, userAction,audioParams ).then(function (response) {
> 			console.log("Audio successfully loaded.")
> 			console.log( audioParams )
> 		}).catch(function (error) {
> 			console.warn("Audio problem:", error);
> 		});
> 	}
> }).catch(function (error) {
> 	console.warn("Audio problem:", error);
> });
> ```
> 
> And it works. Using the `%%MIDI program` in the abc string changes the instrument.

 Thanks ehsansh. Following your advice, I checked and found  mp3 files were not correctly loaded and buffered. Console always show:"Failed to load resource:net::ERR_FAILED    …/…/xxx. mp3" and "Didn't load note acoustic_grand_piano xxx error". 
I don’t know the reason, whether it is sf2 or mp3.js format, I sincerely hope that Paul can provide a successful example .

**Comment by paulrosen** [comment]
Is there a space between the period and mp3 in that error message above or is that just the way it was pasted in this issue? I try to take the path and `curl` it to see if it is actually available if something mysterious happens.

**Comment by ehsansh** [comment]
@watersoft123 I have seen the same error. As Paul mentioned I tried to open that link of the mp3 file and found it does not exist because the `soundFontUrl` was not correct.

**Comment by watersoft123** [comment]
@ehsansh It is not the address error but the cross-domain request is forbidden, maybe for security reasons, js does not allow access to local file. I think we can use js file, just like acoustic_grand_piano-mp3.js, instead of accessing from online,  load and buffer mp3 files from coustic_grand_piano-mp3.js . I hope Paul adds this loading method in the options so that offline playback can be realized.

**Comment by paulrosen** [comment]
The CORS problem is likely with however you are hosting the soundfont. I don't have that problem when I load locally. What kind of server are you using? Does your soundfont URL contain `localhost`?

**Comment by watersoft123** [comment]
> The CORS problem is likely with however you are hosting the soundfont. I don't have that problem when I load locally. What kind of server are you using? Does your soundfont URL contain `localhost`?

Yes,I run it in the browser by double clicking the html file.I use localhost and turn on CORS extension with chrome，but it still doesn't work. I have learned html programming for a short time, and have limited understanding of many things.
![url](https://user-images.githubusercontent.com/57562864/114648317-2b379580-9d11-11eb-8b64-3c1eb1324296.jpg)
![cors](https://user-images.githubusercontent.com/57562864/114648318-2b379580-9d11-11eb-897c-563a0a22cccc.jpg)
![error](https://user-images.githubusercontent.com/57562864/114648316-2b379580-9d11-11eb-9e43-f5f6c1cac99a.jpg)

Can you provide an example? Thank you in advance, my email: tjwater@hotmail.com

**Comment by watersoft123** [comment]
Hi Paul, using XHR to request local files seems to always encounter CORS problems. Is it possible to add a conditional judgment to abcjs-basic.js,and use another method when reading from local. My purpose is to use webview to load this html file in Android,but after testing,I found that I also encounter CORS problems.

**Comment by paulrosen** [comment]
The setting to allow CORS will be in whatever server is powering localhost:8000. CORS is a pain to get right.

You could also put your sound font directly in your project so you get to it with a relative path.

**Comment by paulrosen** [comment]
The seek using seconds is fixed in 6.0.0-beta.33

For the CORS problem, I don't know. I'd have to debug. The easiest thing it to just put the soundfont in your project.
